flask/common/src/main/java/net/corda/flask/common/ManifestEscape.java

```diff
@@ -0,0 +1,89 @@
+package net.corda.flask.common;
+
+import java.text.CharacterIterator;
+import java.text.StringCharacterIterator;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ManifestEscape {
```

Comment:
the problem is that I need to store in the manifest a list of strings, like the list of JVM parameters, and JVM parameters can contain arbitrary characters, like whitespaces or any thing else
e.g
```
java -DsomeProperty="some value" -DsomeOtherProperty="some other value"
```
if I simply stored in the manifest the string
```
-DsomeProperty="some value" -DsomeOtherProperty="some other value"
```
it would be impossible to split the arguments correctly (I could use the double quotes, but property Java property values can contain double quotes too)

Reply:
Then why store them in the Manifest at all? It sounds like a line-oriented text file would work better.

How do we solve this problem at the moment? (Although I fear the answer is "Typesafe Config"... :scream:)

Reply:
newlines are also legal characters in Java properties values, so line-oriented text files wouldn't help

Reply:
> newlines are also legal characters in Java properties values, so line-oriented text files won't help

How about a line-oriented text file with a `.properties` extension? I somehow think that Java _must_ be capable of reading any legal Java property from one of its own properties files, so there's no need to reinvent this wheel. See:
```java
java.util.Properties.load(InputStream);
java.util.Properties.store(OutputStream, String);
```

Reply:
it is not only about JVM properties, any kind of JVM argument can contain spaces, newline, quotes and double quotes, take
```
java -version:"1.6.0_13 1.6*&1.6.0_10+"
```
this is legal and documented in 
```
man java
```

Reply:
And _**realistically**_, exactly how many of those JVM arguments does Corda use? For reference, here is the line from `node/capsule/build.gradle`:
```
jvmArgs = ['-Xmx512m', '-XX:+UseG1GC', '-Djdk.attach.allowAttachSelf=true']
```
It seems that Capsule is quite happy to use `List<String>` here. Although we could also do what Gradle does, i.e. create a single `jvmargs` property:
```
org.gradle.jvmargs=-XX:+UseG1GC -Xmx1g
```

Reply:
FYI the current capsule plugin is broken and doesn't handle this problem at all
```
jvmArgs = ["-DsomeProperty=\"some value\""]
```
breaks the generated executable jar

Reply:
creating a property file with the structure
```
net.corda.flask.jvmarg.1=-XX:+UseG1GC
net.corda.flask.jvmarg.2=-Xmx1g
net.corda.flask.jvmarg.3=...
...
```
would work, but it would be an hack and it would be hard to understand and manually edit.
with the current approach, if you don't use nasty argument the generated string is what you would naturally expect, for example for the corda.jar it would simply be
```
JVM-Args: -Xmx512m -XX:+UseG1GC -Djdk.attach.allowAttachSelf=true
```
while if you use more exotic options, all that happens is that they are just correctly escaped as needed, which, I don't know how else to put it, is simply the right thing to do

Reply:
> FYI the current capsule plugin is broken and doesn't handle this problem at all
> ```
> jvmArgs = ["-DsomeProperty=\"some value\""]
> ```

Or in other words, you have introduced a lot of extra complexity to fix a problem that we do not have.

Reply:
How is
```
JVM-Args: -Xmx512m -XX:+UseG1GC -Djdk.attach.allowAttachSelf=true
```
different from a single property
```
net.corda.flask.jvmargs=-Xmx512m -XX:+UseG1GC -Djdk.attach.allowAttachSelf=true
```
which _can_ be added to a standard property file without all of your custom parsing and escaping code?

Reply:
> Or in other words, you have introduced a lot of extra complexity to fix a problem that we do not have.

Have you ever looked into capsule own code? I'm sure that will change your definition of **"a lot of extra complexity"**
And btw my *extra complexity* fixes a bug that we currently have in all of the programs we ship, you may say that you don't care and nobody ever tried to use a JVM argument containing a space, until the day somebody needs to do it and our code blows up because it's broken

Reply:
> net.corda.flask.jvmargs=-Xmx512m -XX:+UseG1GC -Djdk.attach.allowAttachSelf=true

I'm not sure you've understood that this is **FAULTY**, JVM arguments cannot be split on spaces!

Reply:
Firstly, it doesn't matter how over-engineered Capsule may be because we are getting rid of it.
Secondly, Corda already has a way of setting system property values (including spaces) via `node.conf`. We do not ask our customers to configure them using command line arguments, and so the sky is not going to fall.
And thirdly, I have just discovered that Java 11 supports [argument files](https://docs.oracle.com/en/java/javase/13/docs/specs/man/java.html#java-command-line-argument-files). E.g.
```
$ java @args TestClass
```
where `args` is a line-oriented text file:
```
-Dorg.testing='The cat says "Wibble!"'
-Xmx512m
-XX:+UseG1GC
```

Reply:
another reason for not using a property file is that you still need some form of escaping for arguments forwarded to the child process via
```
java -Dnet.corda.flask.jvm.args="-Dsome.property=\"some\ value\"" -jar corda.jar
```

Reply:
I know about argument files, I use them in `net.corda.flask.launcher.JavaProcessBuilder` to work around the command line length limit, what are you trying to suggest?

Reply:
btw argument files still need escaping (and I do escape them)

Reply:
It should be glaringly obvious what I am suggesting:

_We don't need a lot of custom code implementing functionality that Java already has._

This thread is about a class called `ManifestEscape`, which is apparently for escaping and unescaping attribute values read from `MANIFEST.MF`. And I have spent the better part of a day trying to suggest _**better**_ places than `MANIFEST.MF` for this information. The point about both properties files and argument files is that they can be embedded directly into the flask's `META-INF` directory, with their contents already in the appropriate format. System properties can be read from a standard properties file and mapped to a `List<String>` of either `-Dkey=value` or `-Dkey="value"` arguments, depending on whether or not `value` contains any spaces. The `Properties.load()` method will unescape the properties for you automatically. Whereas an arguments file could either be read into another `List<String>` or dropped to disk and executed "as is" by `@filename` without needing to be unescaped at all.

I cannot for the life of me fathom why you want to put `JVM-Args` in the manifest.

Reply:
Firstly, while it's good to see that people are passionate about getting a good solution, let's be careful that the discourse doesn't get too personal.

In general I think trying to manually escape strings is likely to be a source of bugs. As a result I'd favour a solution that avoids us having to do this, for which there are suggestions on this thread already. It's also worth keeping in mind that the goal is not to replace capsule - it is to replace capsule for our use cases. We can afford to bin complex code if something simpler will suffice for what we want to achieve. I believe this is the case here.